### PR TITLE
Enable mouse wheel input

### PR DESCRIPTION
**Pull request by PSteinhaus**
This change simply adds the funcionality to read mouse wheel input, which (at least for me) is a very necessary feature. My solution is clearly not perfect, as I do rely on somewhat of a workaround to reset the scrolling value to 0 while no scroll events happen. That should probably be changed, but I hope  you forgive me for that, as I am absolutely new to go and haven't been able to find an alternative to this yet, though it's probably pretty easy.

### CODE REVIEW

File: internal/input/input_glfw.go [hunk 1]
```diff
@@ -22,13 +22,16 @@ package input
...
+	"time"
 
 	glfw "github.com/go-gl/glfw/v3.2/glfw"
 )
 
 type Input struct {
 	keyPressed         map[glfw.Key]bool
 	mouseButtonPressed map[glfw.MouseButton]bool
+	scrollY			   float64
+	scrollTime		   time.Time
```
Comment (hajimehoshi) [inline internal/input/input_glfw.go]: Please run `gofmt -s -w .` or the CI test doesn't pass.

File: input.go [hunk 2]
```diff
@@ -52,6 +52,11 @@ func CursorPosition() (x, y int) {
 	return ui.AdjustedCursorPosition()
 }
 
+// MouseScroll returns the y offset of the mouse wheel
```
Comment (hajimehoshi) [inline input.go]: From the comment, I couldn't understand when this function return 0 or non-0 value.

File: input.go [hunk 3]
```diff
@@ -52,6 +52,11 @@ func CursorPosition() (x, y int) {
...
+func MouseScroll() float64 {
+	return input.Get().MouseScroll()
```
Comment (hajimehoshi) [inline input.go]: You should implement this function not only at `input_glfw.go` but also `input_js.go` and `input_mobile.go`. I think it should be possible to implement this for browsers. For mobiles, just return 0 is fine.

File: input.go (hunk 2 above)
Comment (hajimehoshi) [reply input.go]: BTW, is Y axis enough? Doesn't wheeling happen both on X an Y axes?

Comment (PSteinhaus) [reply input.go]: I see your point. I oriented that comment on the variable names in the callback (xoff, yoff). In my mind it works that way that the offset of a normal mouse wheel is always set back to 0 when it's not used. I'll add that.

For me, Y axis was enough, but i guess I should add it while I'm at it.

File: input.go (hunk 3 above)
Comment (PSteinhaus) [reply input.go]: Ok, will do!

Comment (PSteinhaus) [reply input.go]: I implemented the function at `input_mobile.go`  but failed to do so at `input_js.go`. Sry, I'm too inexperienced to see through that.

File: internal/input/input_glfw.go (hunk 1 above)
Comment (PSteinhaus) [reply internal/input/input_glfw.go]: Thx, that solved that problem.

**Comment by PSteinhaus** [comment]
I am sorry, I have no idea why the CI test fails. Do you have any?

**Comment by PSteinhaus** [comment]
I added a `MouseScroll()` function to `input_js.go` , but it's disfunctional and I don't have the knowledge necessary to implement a working version, sorry.
Tell me if you're ok with that. I will of course close this pull request if not.

**Comment by hajimehoshi** [comment]
No problem: could you reave a TODO comment at `input_js.go`? I'll take a look later.

File: input.go [hunk 4]
```diff
@@ -52,6 +52,12 @@ func CursorPosition() (x, y int) {
 	return ui.AdjustedCursorPosition()
 }
 
+// MouseScroll returns the x and y offset of the scroll wheel or scrolling device.
+// For a scroll wheel it returns 0 if the wheel isn't being rolled.
+func MouseScroll() (xoff, yoff float64) {
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply input.go]: As far as I know, GLFW 3.2 doesn't accept touches. So just renaming should be fine :-)

File: internal/input/input_glfw.go [hunk 5]
```diff
@@ -101,6 +111,17 @@ func (i *Input) Update(window *glfw.Window, scale float64) {
 				i.m.Unlock()
 			}
 		})
+		// I added this in here so that it would only be run once too
+		window.SetScrollCallback(func(w *glfw.Window, xoff float64, yoff float64) {
+			i.m.Lock()
+			i.scrollX = xoff
+			i.scrollY = yoff
+			i.scrollTime = time.Now() // this is a rather inelegant way of setting the scroll value to 0 if no scrolling is happening
```
[the remark this answers is not in the document]
Comment (PSteinhaus) [reply internal/input/input_glfw.go]: Thanks for the tip, it's changed now.

File: input.go [hunk 6]
```diff
@@ -52,6 +52,12 @@ func CursorPosition() (x, y int) {
 	return ui.AdjustedCursorPosition()
 }
 
+// MouseScroll returns the x and y offset of the scroll wheel.
+// It returns 0 if the wheel isn't being rolled.
+func MouseScroll() (xoff, yoff float64) {
```
Comment (hajimehoshi) [inline input.go]: Rename this to `MouseWheel`

Comment (PSteinhaus) [reply input.go]: Agreed, that's more fitting. Done.

File: internal/input/input_js.go [hunk 7]
```diff
@@ -91,6 +96,11 @@ func (i *Input) IsMouseButtonPressed(button MouseButton) bool {
 	return false
 }
 
+// TODO: Mouse scroll functionality is not yet implemented in js
+func (i *Input) MouseWheel() (xoff, yoff float64) {
```
Comment (hajimehoshi) [inline internal/input/input_js.go]: Define the counterpart in `input_mobile.go`

File: internal/input/input_js.go [hunk 8]
```diff
@@ -49,6 +49,11 @@ func (i *Input) ClearRuneBuffer() {
 	i.runeBuffer = nil
 }
 
+func (i *Input) ResetScrollValues() {
```
Comment (hajimehoshi) [inline internal/input/input_js.go]: I think `ResetScrollValues` is not needed in `input_js.go`

File: internal/input/input_js.go (hunk 7 above)
Comment (PSteinhaus) [reply internal/input/input_js.go]: Done.

File: internal/input/input_js.go [hunk 9]
```diff
@@ -91,6 +91,11 @@ func (i *Input) IsMouseButtonPressed(button MouseButton) bool {
 	return false
 }
 
+// TODO: Mouse scroll functionality is not yet implemented in js
```
Comment (hajimehoshi) [inline internal/input/input_js.go]: Move this TODO comment in the function body

File: internal/input/input_mobile.go [hunk 10]
```diff
@@ -36,6 +38,11 @@ func (i *Input) IsKeyPressed(key Key) bool {
 	return false
 }
 
+// TODO: Scrolling is not yet implemented on mobile.
```
Comment (hajimehoshi) [inline internal/input/input_mobile.go]: Remove this comment: I don't think I can implement this for mobiles.

File: internal/input/input_js.go (hunk 8 above)
Comment (PSteinhaus) [reply internal/input/input_js.go]: True, I'll add a TODO comment in `ui_js.go`

File: internal/input/input_glfw.go [hunk 11]
```diff
@@ -101,6 +115,13 @@ func (i *Input) Update(window *glfw.Window, scale float64) {
 				i.m.Unlock()
 			}
 		})
+		// I added this in here so that it would only be run once too
+		window.SetScrollCallback(func(w *glfw.Window, xoff float64, yoff float64) {
```
[the remark this answers is not in the document]
Comment (PSteinhaus) [reply internal/input/input_glfw.go]: But using the method with `i.callbacksInitialized` makes it easier to understand and maintain the code, so it's probably better, I agree.

**Comment by hajimehoshi** [comment]
Thank you for using and contributing to my lib! Detecting mouse wheel state sounds an important feature that I didn't realize before you created this PR :-)

**Pull request merged**